File: server/src/services/duration.ts
import type { PostgresJsDatabase } from 'drizzle-orm/postgres-js';
import { db, type IBaseService } from '.';
import { duration, task } from '@/schema';
import { InternalError, ServiceSuccess } from '@/enums';
import { InternalServerError } from '@/errors';
import { eq } from 'drizzle-orm';

interface IDurationService extends IBaseService<TDuration, number> {}

class DurationService implements IDurationService {
    db: PostgresJsDatabase<Record<string, number>>;

    constructor() {
        this.db = db();
    }

    async list(): Promise<TDuration[]> {
        throw new Error('Not implemented');
    }

    async create(value: Omit<TDuration, 'id'>): Promise<{ insertedId: number }> {
        const result = await this.db
            .insert(duration)
            .values(value)
            .returning({ insertedId: duration.id });
        if (result.empty()) throw new InternalServerError(InternalError('duration').CREATE);
        return result[0];
    }

    async update(): Promise<string> {
        throw new Error('Not implemented');
    }

    async remove(): Promise<string> {
        throw new Error('Not implemented');
    }

    async logsFromNode(nodeId: number): Promise<TDurationFromNode[]> {
        return await this.db
            .select({
                id: duration.id,
                taskTag: {
                    id: task.id,
                    title: task.title,
                    color: task.color,
                },
                description: duration.description,
                timeOnTask: duration.timeOnTask,
                startFrom: duration.startFrom,
                endAt: duration.endAt,
            })
            .from(duration)
            .innerJoin(task, eq(duration.taskId, task.id))
            .where(eq(task.nodeId, nodeId));
    }
}

export function useDuration() {
    return new DurationService();
}

File: server/internal/services/duration.go
package services

import (
	success "sukejuuru/internal/enums"
	"sukejuuru/internal/models"
	"sukejuuru/internal/utils"

	"github.com/jmoiron/sqlx"
)

type durationService struct {
	db *sqlx.DB
}

func UseDuration() *durationService {
	return &durationService{db: utils.UsePostgres()}
}

func (service durationService) FindAllByNode(nodeId int) ([]models.Log, error) {
	var logs []models.Log
	query := `
		SELECT d.id,
			t.id, t.title, t.color
			d.description, d.time_on_task, d.start_from, d.end_at
		FROM duration d
		JOIN task t ON d.task_id = t.id
		WHERE t.node_id = $1
		ORDER BY d.id DESC
	`

	if err := service.db.Select(&logs, query, nodeId); err != nil {
		return nil, err
	}
	return logs, nil
}

func (service durationService) Create(values models.Duration) (string, error) {
	query := `
		INSERT INTO duration (time_on_task, start_from, end_at, description, task_id)
		VALUES (:time_on_task, :start_from, :end_at, :description, :task_id)
	`
	if _, err := service.db.NamedExec(query, values); err != nil {
		return "", err
	}
	return success.CREATED, nil
}
